csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,261 @@
+    } else if (abs_id.is<ValGroupAndItsGraph>()) {
+      if (graph_ == nullptr) {
+        graph_ = abs_id.as<ValGroupAndItsGraph>().graph;
+      } else {
+        NVF_ERROR(graph_ == abs_id.as<ValGroupAndItsGraph>().graph);
+      }
+      return abs_id.as<ValGroupAndItsGraph>().group;
+    }
+    NVF_ERROR(false, "AbstractId must be IterDomain* or ValGroupAndItsGraph");
+  }
+
+  //! Work backward from each loop IterDomain in concrete. When we find an
+  //! IterDomain* with a ValGroup in scheduled_val_groups, we map to it
+  //! (including siblings) in concrete_ids_. Assert that we do find a
+  //! scheduled ValGroup, otherwise we would not know where to place this
+  //! IterDomain.
+  void findNearestProducers() {
+    // Record all ValGroups that are producers of those in abstract_. This lets
+    // us find starting IterDomains for scheduling concrete's loop domain.
+    // TODO: this should probably use a new more general utility that resembles
+    // StmtSort::getExprsTo in val_graph_visitor.h
+    std::unordered_set<ValGroup> scheduled_val_groups;
+    std::stack<ValGroup> vg_stack;
+    for (const AbstractId& abs_id : abstract_.domain) {
+      vg_stack.push(abstractIdToValGroup(abs_id));
+    }
+    while (!vg_stack.empty()) {
+      ValGroup vg = vg_stack.top();
+      vg_stack.pop();
+      bool inserted = scheduled_val_groups.insert(vg).second;
+      if (!inserted) {
+        // Avoid cycles
+        continue;
+      }
+      for (const ExprGroup& eg : graph_->getDefinitions(vg)) {
+        for (Expr* e : *eg) {
+          for (Val* inp : e->inputs()) {
+            ValGroup vg_inp = graph_->toGroup(inp);
+            vg_stack.push(vg_inp);
+          }
+        }
+      }
+    }
+
+    // Now traverse c2p from concrete loop domain, stopping when we find a
+    // scheduled ValGroup
+    std::stack<IterDomain*> id_stack;
+    for (IterDomain* id : concrete_->getLoopDomain()) {
+      id_stack.push(id);
+    }
+    while (!id_stack.empty()) {
+      IterDomain* id = id_stack.top();
+      id_stack.pop();
+      ValGroup g = graph_->toGroup(id);
+      if (scheduled_val_groups.find(g) != scheduled_val_groups.end()) {
+        concrete_ids_.emplace(g, id);
+        continue;
+      }
+      NVF_ERROR(
+          id->definition() != nullptr,
+          "Root IterDomain ",
+          id->toString(),
+          " does not appear in the history of any ValGroups in abstract tensor");
+      for (Val* inp : id->definition()->inputs()) {
+        if (auto inp_id = dynamic_cast<IterDomain*>(inp)) {
+          id_stack.push(inp_id);
+        }
+      }
+    }
+  }
+
+  IterDomain* replayAbstractId(AbstractId abs_id) {
+    ValGroup g = abstractIdToValGroup(abs_id);
+
+    // This holds ValGroups that we cannot compute from concrete_'s root. When
+    // we detect that any inputs of an ExprGroup are in these groups, we know
+    // that we cannot compute that ExprGroup so we should skip it.
```

Comment:
Could you give an example in the comment?

Reply:
Added